include/drivers/gpio.h

```diff
@@ -193,17 +544,416 @@ static inline int z_impl_gpio_disable_callback(struct device *port,
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Get logical level of an input pin.
+ *
+ * Get logical level of an input pin taking into account GPIO_ACTIVE_LOW flag.
+ * If pin is configured as Active High, a low physical level will be interpreted
+ * as logical value 0. If pin is configured as Active Low, a low physical level
+ * will be interpreted as logical value 1.
+ *
+ * Note: If pin is configured as Active High, the default, gpio_pin_get()
+ *       function is equivalent to gpio_pin_get_raw().
+ *
+ * @param port Pointer to the device structure for the driver instance.
+ * @param pin Pin number.
+ *
+ * @retval 1 If pin logical value is 1 / active.
+ * @retval 0 If pin logical value is 0 / inactive.
+ * @retval -EIO I/O error when accessing an external GPIO chip.
+ */
+static inline int gpio_pin_get(struct device *port, unsigned int pin)
+{
+	gpio_port_value_t value;
+	int ret;
+
+	__ASSERT(pin < GPIO_MAX_PINS_PER_PORT, "Invalid pin number");
+
+	ret = gpio_port_get(port, &value);
+	if (ret == 0) {
+		ret = (value & BIT(pin)) != 0 ? 1 : 0;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Set physical level of an output pin.
+ *
+ * Writing value 0 to the pin will set it to a low physical level. Writing any
+ * value other than 0 will set it to a high physical level. This function
+ * ignores GPIO_ACTIVE_LOW flag.
+ *
+ * @param port Pointer to the device structure for the driver instance.
+ * @param pin Pin number.
+ * @param value Value assigned to the pin.
+ *
+ * @retval 0 If successful.
+ * @retval -EIO I/O error when accessing an external GPIO chip.
+ */
+static inline int gpio_pin_set_raw(struct device *port, unsigned int pin,
+				   int value)
+{
+	int ret;
+
+	__ASSERT(pin < GPIO_MAX_PINS_PER_PORT, "Invalid pin number");
+
+	if (value != 0)	{
+		ret = gpio_port_set_bits_raw(port, BIT(pin));
+	} else {
+		ret = gpio_port_clear_bits_raw(port, BIT(pin));
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Set logical level of an output pin.
+ *
+ * Set logical level of an output pin taking into account GPIO_ACTIVE_LOW flag.
+ * Value 0 sets the pin in logical 0 / inactive state. Any value other than 0
+ * sets the pin in logical 1 / active state. If pin is configured as Active
+ * High, the default, setting it in inactive state will force the pin to a low
+ * physical level. If pin is configured as Active Low, setting it in inactive
+ * state will force the pin to a high physical level.
+ *
+ * Note: If pin is configured as Active High, gpio_pin_set() function is
+ *       equivalent to gpio_pin_set_raw().
+ *
+ * @param port Pointer to the device structure for the driver instance.
+ * @param pin Pin number.
+ * @param value Value assigned to the pin.
+ *
+ * @retval 0 If successful.
+ * @retval -EIO I/O error when accessing an external GPIO chip.
+ */
+static inline int gpio_pin_set(struct device *port, unsigned int pin, int value)
+{
+	struct gpio_driver_data *const data =
+			(struct gpio_driver_data *const)port->driver_data;
```

Comment:
You're making an immutable pointer to mutable data, which isn't very useful and looks weird because other pointers aren't treated that way.  The fixup I provided made clear that the code should not be changing `invert`:
```
-       struct gpio_driver_data *const data = port->driver_data;
+       const struct gpio_driver_data *data =
+               (const struct gpio_driver_data *)port->driver_data;
```
3 occurrences if you want to change it.